src/libraries/Microsoft.Extensions.Options/src/ValidateOptionsResultBuilder.cs

```diff
@@ -0,0 +1,126 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Diagnostics;
+using System.Linq;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.Extensions.Options
+{
+    /// <summary>
+    /// Builds <see cref="ValidateOptionsResult"/> with support for multiple error messages.
+    /// </summary>
+    [DebuggerDisplay("{_errors?.Count} errors")]
+    public class ValidateOptionsResultBuilder
+    {
+        private const string MemberSeparatorString = ", ";
+
+        private List<string>? _errors;
+
+        /// <summary>
+        /// Creates new instance of the <see cref="ValidateOptionsResultBuilder"/> class.
+        /// </summary>
+        public ValidateOptionsResultBuilder() { }
+
+        /// <summary>
+        /// Adds validation error.
+        /// </summary>
+        /// <param name="error">Content of error message.</param>
+        /// <param name="propertyName">THe property in the option object which contains an error.</param>
+        public void AddError(string error, string? propertyName = null)
+        {
+            ThrowHelper.ThrowIfNull(error);
+            Errors.Add(propertyName is null ? error : $"Property {propertyName}: {error}");
+        }
+
+        /// <summary>
+        /// Adds any validation error carried by the <see cref="ValidationResult"/> instance to this instance.
+        /// </summary>
+        /// <param name="result">The instance to consume the errors from.</param>
+        public void AddResult(ValidationResult? result)
+        {
+            if (result?.ErrorMessage is not null)
```

Comment:
If a `null` result is success, what does it mean for result to be non-null but ErrorMessage to then be null?

Reply:
https://github.com/dotnet/runtime/blob/99547ae0bdb458a4f147167944a298ee2d7486b7/src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationResult.cs#L37